django/db/models/query.py

```diff
@@ -815,6 +815,25 @@ def complex_filter(self, filter_obj):
         else:
             return self._filter_or_exclude(None, **filter_obj)
 
+    def _combinator_query(self, combinator, *other_qs):
+        # Clone the query so we inherit the select list and everything
+        clone = self._clone()
+        # Clear limits and ordering, so they can be reapplied
+        clone.query.clear_ordering(True)
+        clone.query.clear_limits()
+        clone.query.combined_queries = (self.query,) + tuple(qs.query for qs in other_qs)
+        clone.query.combinator = combinator
+        return CombinedQuerySet(clone)
+
+    def union(self, *other_qs):
```

Comment:
The thing is that even if the ORM doesn't have support for it yet using `distinct()` to implement `(UNION|INTERSECT) ALL` might prevent us from adding proper support in the future.

What I suggest doing here is setting `query.combinator.all = kwargs['all']` and preventing using `distinct()` on `CombinedQuerySet`.

The difference between ordering and combination operation is that the former operates on the _combined_ set of rows while the latter operates on how these rows are combined. I would suggest that options related to combination be passed as `kwargs` (such as `all`) and actions operating of the combined result (`CombinedQuerySet` instances) be added as methods (`order_by`, _slicing_).